File: semester3/advanced-programming-methods/ToySocialNetwork/java/application/utils/pair/Pair.java
package application.utils.pair;

/**
 * The generic class <b>Pair</b> contains 2 objects.
 **/
public class Pair<E1, E2> {
    private final E1 first;
    private final E2 second;

    public Pair (E1 first, E2 second) {
        this.first = first;
        this.second = second;
    }

    public E1 first() {
        return first;
    }

    public E2 second() {
        return second;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Pair<?, ?> pair = (Pair<?, ?>) o;

        return first.equals(pair.first) && second.equals(pair.second);
    }
}

File: 3rd Semester/Advanced Programming Methods/Homework/JrNBALeagueRomania/models/Pair.cs
namespace JrNBALeagueRomania.models
{
    internal class Pair<T1, T2>
    {
        public T1 First { get; set; }
        public T2 Second { get; set; }

        public Pair(T1 first, T2 second)
        {
            First = first;
            Second = second;
        }

        public override bool Equals(object? obj)
        {
            if (obj == null || obj.GetType() == GetType())
            {
                return false;
            }
            Pair<T1, T2> o = (Pair<T1, T2>) obj;
            if (First == null)
            {
                if (Second == null)
                {
                    return o.First == null && o.Second == null;
                }
                return o.First == null && Second.Equals(o.Second);
            }
            if (Second == null)
            {
                return o.Second == null && First.Equals(o.First);
            }
            return First.Equals(o.First) && Second.Equals(o.Second);
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(First, Second);
        }
    }
}
